### PR TITLE
Refactor `TinyList::contains` and `len` to iterate instead of recurse

### PR DESCRIPTION
**Pull request by llogiq**

### CODE REVIEW

**Comment by Alexendoo** [comment]
Ping from triage, any updates? @eddyb

**Comment by llogiq** [comment]
Rebased.

File: src/librustc_data_structures/tiny_list/tests.rs [hunk 1]
```diff
@@ -111,6 +111,27 @@ fn bench_insert_one(b: &mut Bencher) {
     })
 }
 
+#[bench]
+fn bench_contains_empty(b: &mut Bencher) {
+    b.iter(|| {
+        TinyList::new().contains(&1)
```
Comment (eddyb) [inline src/librustc_data_structures/tiny_list/tests.rs]: You need to `black_box` the list, here and in the other tests, so the benchmark is actually doing anything.

**Review by eddyb — APPROVED** [review]
r=me with benches fixed

File: src/librustc_data_structures/tiny_list/tests.rs (hunk 1 above)
Comment (llogiq) [reply src/librustc_data_structures/tiny_list/tests.rs]: Added.

File: src/librustc_data_structures/tiny_list/tests.rs [hunk 2]
```diff
@@ -98,36 +98,57 @@ fn test_remove_single() {
...
         list.insert(1);
     })
 }
 
 #[bench]
 fn bench_insert_one(b: &mut Bencher) {
     b.iter(|| {
-        let mut list = TinyList::new_single(0);
+        let mut list = black_box(TinyList::new_single(0));
         list.insert(1);
```
Comment (eddyb) [inline src/librustc_data_structures/tiny_list/tests.rs]: Oh you also need to produce `list` from this block, so the insert is not optimized out.

File: src/librustc_data_structures/tiny_list/tests.rs [hunk 3]
```diff
@@ -98,36 +98,57 @@ fn test_remove_single() {
 #[bench]
 fn bench_insert_empty(b: &mut Bencher) {
     b.iter(|| {
-        let mut list = TinyList::new();
+        let mut list = black_box(TinyList::new());
         list.insert(1);
```
Comment (eddyb) [inline src/librustc_data_structures/tiny_list/tests.rs]: Same here.

**Comment by llogiq** [comment]
@eddyb r?

**Comment by JohnCSimon** [comment]
Ping from triage 
@llogiq @eddyb I'm not sure why this is in s-waiting-on-author instead of s-waiting-on-review 
what is the state of this PR? 
thank you!

**Comment by llogiq** [comment]
I made the changes to the benchmarks as requested. @eddyb please clarify what is missing.

**Comment by eddyb** [comment]
@llogiq Oops, you didn't mark the comments as resolved and GitHub didn't say "outdated" because you just added a line afterwards... So I thought you hadn't done those changes yet, my bad.

@bors r+

**Pull request merged**